test/test_session.py

```diff
@@ -173,6 +175,46 @@ def _test_ops(self, client, *ops):
                         "%s did not return implicit session to pool" % (f.__name__,),
                     )
 
+    def test_implicit_sessions_checkout(self):
+        # "To confirm that implicit sessions only allocate their server session after a
+        # successful connection checkout" test from Driver Sessions Spec.
+        succeeded = False
+        for _ in range(5):
+            listener = EventListener()
+            client = rs_or_single_client(
+                event_listeners=[listener], maxPoolSize=1, retryWrites=True
+            )
+            cursor = client.db.test.find({})
+            ops: List[Tuple[Callable, List[Any]]] = [
+                (client.db.test.find_one, [{"_id": 1}]),
+                (client.db.test.delete_one, [{}]),
+                (client.db.test.update_one, [{}, {"$set": {"x": 2}}]),
+                (client.db.test.bulk_write, [[UpdateOne({}, {"$set": {"x": 2}})]]),
+                (client.db.test.find_one_and_delete, [{}]),
+                (client.db.test.find_one_and_update, [{}, {"$set": {"x": 1}}]),
+                (client.db.test.find_one_and_replace, [{}, {}]),
+                (client.db.test.aggregate, [[{"$set": {"x": 1}}]]),
+                (client.db.test.find, [{}]),
+                (client.server_info, [{}]),
+                (client.db.aggregate, [[{"$listLocalSessions": {"allUsers": True}}]]),
+                (client.db.list_collections, []),
+                (cursor.distinct, ["_id"]),
+            ]
+            threads = []
+            listener.results.clear()
+            for op, args in ops:
+                threads.append(threading.Thread(target=op, args=args))
+                threads[-1].start()
+            for thread in threads:
+                thread.join()
```

Comment:
There's one more hidden problem with this test. The test does not assert that the threads actually succeeded in running their operation. Change this to use ExceptionCatchingThread and then assert that `exc` is None for each thread.

Reply:
I changed it to use exception catching thread and it is still failing on Github Actions.

Reply:
Nevermind, I see what you mean. Will fix.